### ISSUE TITLE
Timezone awareness?

### ISSUE THREAD

**Issue by bradvogel** [lacona/lacona-phrase-datetime#1]
I'm evaluating time strings (such as `3pm`) on my server, which might not be in the same timezone from where the string originates. This can result in incorrect parsing.

Do you have any pointers on how this can be fixed? I'm happy to file a PR.

**Comment by bradvogel** [comment]
In researching this a bit, I think this can be done by setting the `.zone()` on the moment object when `set()`ing times:

So the join function might look like this, where `zone` is passed in as an optional parameter.

```
export function join (date, time, zone) {
  return moment(date).zone(zone).set(time).toDate()
}
```

Does this look right to you?

**Comment by brandonhorst** [comment]
Hmmm - good question - I haven't considered a non-local use case.

`Date`, `DateTime`, `DateRange`, and `Range` all export JS `Date` objects. They use `moment()` under the covers. I *believe* that if you used [Moment Timezone](http://momentjs.com/timezone/docs/#/using-timezones/parsing-in-zone/) and passed the time zone as a a second argument to every `moment()` constructors (in `src/helpers.js`), it would output dates in the correct time zone. I haven't used it before, but it seems like that should work fine. We should allow `Date`, `DateTime`, `DateRange`, and `Range` to accept a `timeZone` prop.

Then your server call could simply modify the grammar with each request to pass in the appropriate timezone. Something like this.

```js
var userTimeZone = /* whatever */

class MyCommand extends Phrase {
  describe () {
    return <DateTime timeZone={this.props.userTimeZone} />
  }
}

parser.grammar = <MyCommand userTimeZone={userTimeZone} />
parser.parseArray()
```

Does that look reasonable?

This isn't a priority for me because [Lacona](http://lacona.io/) only runs locally - could you take a shot at implementing? I know the docs are weak, please shoot any questions my way.

Also, I'd love to hear more about what you are building with Lacona. Dates especially are fairly processing-intensive (though the changes in progress on the https://github.com/lacona/lacona-phrase-datetime/tree/ellipsis branch improves it dramatically), and I suspect you may run into issues at scale, at least in a single-threaded Node.js environment.

**Comment by brandonhorst** [comment]
Yeah, I think that `moment().utcOffset()` would work as well (`moment().zone()`) appears to be deprecated).

**Comment by bradvogel** [comment]
Awesome, filed https://github.com/lacona/lacona-phrase-datetime/pull/2 for your review.

For use case, I'm adding a Mixmax slash command (see sdk.mixmax.com) to easily schedule an event over email (e.g. `/event have dinner at home tomorrow 3-6pm` will create a calendar event). We'll probably add a bunch of accelerators for various email-related tasks. Since the processing is done server-side and the performance bottleneck is network time, I'm not too worried about processing time.

**Comment by brandonhorst** [comment]
Which phrases are you using? Development is going to continue based on the ellipsis branch (which sees simpler code and 10x+ performance improvements), but only DateTime and Range are working properly at the moment. If you only need those phrases, we should make this change to that branch. If you need other phrases, maybe we should get the other praises working in ellipsis.

**Comment by bradvogel** [comment]
I only need the Range type for now. But it looks like Range uses all the other types implicitly - right? What isn't supported in the ellipsis branch currently? Let me know and I'll be happy to port this fix over to that branch. If the ellipsis branch won't be ready soon, I'd appreciate if you could help me fix the three broken unit tests in the PR. I'm sure you can spot the bug in a minute or two :)

Thanks again for this wonderful framework!

**Comment by brandonhorst** [comment]
I've just merged Ellipsis into master. `DateRange` and `TimeRange` are not working, but I'm going to consider those non-essential. ([Lacona](http://www.lacona.io/) actually only uses `DateTime` and `Range` for now).

I'm going to look into adding your timezone changes to master now. Sorry for the huge delay in this, got caught up with a lots of LaconaApp things!

**Comment by bradvogel** [comment]
Brandon, no problem, and awesome work on new lacona.io.

If this PR adds too much code bloat, I'd be glad to make it a separate Lacona plugin (perhaps `lacona-phrase-datetime-with-timezone`?). I just need your guidance on unbreaking those few unit tests, then I'm set!

**Comment by brandonhorst** [comment]
So I'm looking at the code from your PR, and I see the problem. 

User inputs `"2/3/2003 at 2pm"`

Lacona parses using `DateWithTimeOfDayAndTime`, ends up with:

```js
{
  date: Date(2/3/2003 0:00:00),
  time: {hour: 14, minute: 0, second: 0}
}
```

It calls `join`. If I combine `join` and `getMoment`, we get:

```js
return moment.tz(Date(2/3/2003 0:00:00), 'Pacific/Honolulu').set({hour: 14, minute: 0, second: 9}).toDate()
```

The issue is that passing a date to `moment.tz` means "take this date in local time and convert it into this new time zone." If I ran this code in Boston, it would reduce to:

```
return moment(Date(2/2/2003 17:00:00)).set({hour: 14, minute: 0, second: 9}).toDate()
```

In SF, it would reduce to

```
return moment(Date(2/2/2003 20:00:00)).set({hour: 14, minute: 0, second: 9}).toDate()
```

If you ran this code in Beijing, it would actually work, because adjusting the time zone wouldn't change the date component.

---

In the case of abolute dates, you need to make sure that you are not converting the date components to new time zones independently, or you will run into problems.

---

This is a tricky problem and I still haven't wrapped my mind around it entirely. To correctly solve this issue, I believe we would need to refactor all of the phrases to use UTC time internally. I think that ultimately, the right thing to do is to fork this repo and make a separate phrase - 25.6k gz for `moment-timezone-all-years` is not good for in-browser applications that don't need time zone logic.

However, for your own purposes, at least in the proof-of-concept phase, you *could* do something that seems a bit kludgey but could work well, at least for now. Because Lacona is syncronous, you could just stub out the time using `lolex` before you process the user's input. Something like this:

```js
var clock = lolex.install(global, clientLocalTime)
var outputs = lacona.parseArray(input)
clock.uninstall()
```

Let me know your thoughts. And sorry again for the delay on this.

**Comment by bradvogel** [comment]
Ah, I see, thanks! Completely overlooked that. I couldn't get the `lolex` solution unfortunately.

**Comment by brandonhorst** [comment]
Closing this for now. I would love to see time zones supported in some way, but it's not a priority right now.

**Comment by brandonhorst** [comment]
Looking into this more. Using lolex doesn't actually work because `date.set*` calls, which `moment` uses under the covers, are not mocked by lolex. I submitted an issue on sinon (lolex's papa) here: https://github.com/sinonjs/lolex/issues/41

Unless we can find a lolex workaround, this wil indeed* require some lacona code changes to fix, and it will likely require passing the timezone offset into the time phrases.

I will need to look into this more.

**Comment by brandonhorst** [comment]
Demo here: https://gist.github.com/brandonhorst/a190336eb2812906e50b

**Comment by brandonhorst** [comment]
Implemented in 0.9.0 (https://github.com/lacona/lacona-phrase-datetime/commit/0b75e395e22c7963f369d98e393f9584044d7c56).

Basically, grab the `new Date().getTimezoneOffset()` from the client's system, and send it in your network request or whatever. Then, sent `<DateTime timezoneOffset={tzOffset} />` in your grammar.

Note that this value is the *negative* number of minutes offset from UTC. So Beijing is UTC+8, so the timezoneOffset is `8 * 60 * -1 = -480`. Who knows why JS devs decided to do it this way.

The output date will be adjusted to behave as though it was input in the given `timezoneOffset`.

Take a look and see if this suits your needs. THanks

**Comment by bradvogel** [comment]
Looks good, thanks!
